### PR TITLE
fix(library/type_context): `?m_1` and `(λ x, ?m_1) y` are definitionally equal

### PR DESCRIPTION
**Pull request by gebner**
Currently, `quick_is_def_eq` incorrectly returns false (i.e., not defeq) for these expressions.

### CODE REVIEW

**Comment by cipher1024** [comment]
Can you summarize why this commit is correct and is the kernel using `quick_is_def_eq`? Also, can you check that mathlib builds with it?

**Comment by gebner** [comment]
> Can you summarize why this commit is correct

The type context uses the same code for unification and checking definitional equality of terms, this function is called `is_def_eq`.  The variables `m_update_left` and `m_update_right` in the `type_context` class control whether we assign metavariables in the left (resp., right) expression.  They hence control whether we do unification or definitional equality checks.  This issue concerns the case when these variables are set to false, i.e., when we perform definitional equality checks.

The `is_def_eq` function is constructed in several layers:
 * `is_def_eq` calls `is_def_eq_core`, and then logs a trace message and handles postponed universe level unification problems
 * `is_def_eq_core` calls `is_def_eq_core_core`, and adds the pair of the expressions to the cache if they have been found to be definitionally equal
 * `is_def_eq_core_core` is the core function.  It does βηιζ-reduction as well as δ-reduction guided by the definitional height heuristic.  Then it tries some ad-hoc heuristics like offset equations and unification hints.  After that, it does the "typical" unification steps: e.g. reduce `f a =?= g b` to `f =?= g` and `a =?= b`.  Finally, it implements proof irrelevance.  However, at the very beginning of `is_def_eq_core_core` (before we reduce anything) is a quick check that should detect unifiable/non-unifiable terms without the costly βδιζ-reduction done in `is_def_eq_core_core`; this function is called `quick_is_def_eq`.
 * `quick_is_def_eq` has some pretty straightforward checks: return true if the pointers are equal, or return true if we have cached the definitional equality.  Then it unfolds metavariables.  If there is a still a metavariable on one side, then it tries to assign it, however the code essentially works like this [comments are my reading of the code]:
```c++
if (f1 is a metavariable) {
  if (m_update_left) {
    // Try to assign metavariable
  } else {
    // The metavariable is not syntactically equal to the other expression.
    // We cannot assign the metavariable, hence they cannot be
    // definitionally equal.
    // ⚡
    return l_false;
  }
}
```
  Alas, this is not correct.  The expressions could still be definitionally equal after reduction.  This PR changes the `l_false` to `l_undef`, thereby proceeding with `is_def_eq_core_core`.

> kernel using quick_is_def_eq

Of course not.  The kernel uses different code for type checking, namely the `type_checker` instead of the `type_context`.

> Also, can you check that mathlib builds with it?

The mathlib build succeeds with debug assertions enabled.
